manager/orchestrator/restart/restart.go

```diff
@@ -237,70 +237,116 @@ func (r *Supervisor) shouldRestart(ctx context.Context, t *api.Task, service *ap
 		log.G(ctx).WithError(err).Error("invalid restart lookback window")
 		return restartInfo.totalRestarts < t.Spec.Restart.MaxAttempts
 	}
-	lookback := time.Now().Add(-window)
+
+	var timestamp time.Time
+	// Prefer the manager's timestamp over the agent's, since manager
+	// clocks are more trustworthy.
+	if t.Status.AppliedAt != nil {
+		timestamp, err = gogotypes.TimestampFromProto(t.Status.AppliedAt)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("invalid task status AppliedAt timestamp")
+			return restartInfo.totalRestarts < t.Spec.Restart.MaxAttempts
+		}
+	} else {
+		// It's safe to call TimeStampFromProto with a nil timestamp
+		timestamp, err = gogotypes.TimestampFromProto(t.Status.Timestamp)
+		if t.Status.Timestamp == nil || err != nil {
+			log.G(ctx).WithError(err).Error("invalid task completion timestamp")
+			return restartInfo.totalRestarts < t.Spec.Restart.MaxAttempts
+		}
+	}
+	lookback := timestamp.Add(-window)
+
+	numRestarts := uint64(restartInfo.restartedInstances.Len())
 
 	var next *list.Element
 	for e := restartInfo.restartedInstances.Front(); e != nil; e = next {
 		next = e.Next()
 
 		if e.Value.(restartedInstance).timestamp.After(lookback) {
+			for e2 := restartInfo.restartedInstances.Back(); e2 != nil; e2 = e2.Prev() {
```

Comment:
> restartInfo.restartedInstances is a doubly-linked list where the front is the oldest restarted instance, and the back is the newest restarted instance?

Correct

> This block does 2 things simultaneously: clearing any recorded history that happened before lookback, and also identifying the number of restarts between lookback and the timestamp for this task?

Yeah, this is a mess.

> If so, it might be slightly easier to read if this were broken into 2 separate loops - one that reaps restart history, and the other that ignores any restarts that happened after the task status timestamp.

I agree. There's no reason for them to be nested. Thanks for the suggestion. I've moved the inner loop after the outer one. This actually revealed a bug, because the wrong iteration variable was being used :(.

> But if it truncates the restart history relative to the timestamp of the task, wouldn't calling ShouldRestart on a later task truncate history that might be needed for an earlier task? Or would that history never be needed again?

That's right that the history would never be needed again. We would only ever make a restart decision for the latest task in a slot.

I just discovered a minor case where we do expect to have the history for older tasks, and I've pushed a fix for it.

> Also, what would cause restartedInstances to have more restarts than the last status update of the task?

`ShouldRestart` gets called on older tasks as well, for example when `updatableAndDeadSlots` calls `IsTaskUpdatable`. That said, I took a closer look at this and the logic is pretty bogus. An older task could get updated by the updater, which doesn't really make sense - it should only consider the most recent task. I've pushed another commit that redoes this. Now `shouldRestart` should only be called on the most recent task. I've left the loop that ignores restarts which happened after the task was restarted, because it seems most correct to have it in place, but it shouldn't be necessary anymore, and I'm fine with removing it.